### PR TITLE
add qtensor support for custom device

### PR DESCRIPTION
**Pull request by heidongxianhua**
Fixes #ISSUE_NUMBER
As QTensor with custom device, like Tensor, we also want to support it. But new a qtensor is different to the existing func for our custom-device, so we add a func so that we can register a func to new qtensor for custom device if needed, and we also could use the existing func if we do not need to  register a func. @albanD  could you have a look?

### CODE REVIEW

File: aten/src/ATen/quantized/Quantizer.cpp [hunk 1]
```diff
@@ -32,6 +32,28 @@ namespace {
...
+void SetNewQTensorImpl(at::DeviceType t, at::NewQTensorFuncType impl) {
+  std::lock_guard<std::mutex> lock(_qtensor_mutex_lock);
+  TORCH_CHECK(t == at::DeviceType::PrivateUse1, "SetNewQTensorImpl only support `PrivateUse1` now.");
+  new_qtensor_impls[static_cast<int>(t)] = impl;
+}
+
+const at::NewQTensorFuncType& GetNewQTensorImpl(at::DeviceType t) {
+  TORCH_CHECK(t == at::DeviceType::PrivateUse1, "GetNewQTensorImpl only support `PrivateUse1` now.");
+  auto& impl = new_qtensor_impls[static_cast<int>(t)];
+  TORCH_CHECK(impl.has_value(), "Please SetNewQTensorImpl first.")
```
Comment (jerryzh168) [inline aten/src/ATen/quantized/Quantizer.cpp:46]: should this happen before L45? and does this need a lock?

File: test/cpp_extensions/open_registration_extension.cpp [hunk 2]
```diff
@@ -42,11 +48,22 @@ void abs_kernel(::at::TensorIteratorBase& iter) {
   abs_counter += 1;
 }
 
+void quantize_tensor_per_tensor_affine_foo(
```
Comment (jerryzh168) [inline test/cpp_extensions/open_registration_extension.cpp:51]: what is this function?

**Comment by jerryzh168** [comment]
I don't think I have enough context to review this, maybe @ezyang?

**Comment by albanD** [comment]
PrivateUse1 is our out-of-tree device extension point. You can think of it as the same as cuda but out of core.

**Comment by ezyang** [comment]
Why are you not able to use the pre-existing QTensorImpl?

**Comment by heidongxianhua** [comment]
> Why are you not able to use the pre-existing QTensorImpl?

@ezyang  yeah, similar to tensor, for our custom-device, there will be some additional format and other information, we inherit TensorImpl to achieve these.  And for Tensor, the creation of Tensor is done through the `empty` operator, which can be dispatched to specific devices through the dispatcher mechanism. So for QTensor, we also want to inherit QTensorImpl, but for QTensor, the creation of QTensor is realized through the api named `new_qtensor` only could used in c++, so we want to extend this `new_qtensor` to achieve a dispacthed-like effect.

**Comment by jerryzh168** [comment]
@heidongxianhua can you describe the full use case? is this related to running a pytorch eager mode quantized model in some custom device?

**Comment by jerryzh168** [comment]
just FYI, new versions of pytorch quantization is planning to move away from QTensor and just use plain pytorch Tensor instead so that we are not tied to pytorch core: https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/fx/_decomposed.py#L56

**Comment by heidongxianhua** [comment]
> just FYI, new versions of pytorch quantization is planning to move away from QTensor and just use plain pytorch Tensor instead so that we are not tied to pytorch core: https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/fx/_decomposed.py#L56

As for QTensor, there are many  dedicated operators such as `quantize_per_tensor` and `quantize_per_tensor_dynamic`, in these operator, pytorch will create `QTensor`. we could use these operators to do model Quantization on our custom device.

**Comment by jerryzh168** [comment]
> > just FYI, new versions of pytorch quantization is planning to move away from QTensor and just use plain pytorch Tensor instead so that we are not tied to pytorch core: https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/fx/_decomposed.py#L56
> 
> As for QTensor, there are many dedicated operators such as `quantize_per_tensor` and `quantize_per_tensor_dynamic`, in these operator, pytorch will create `QTensor`. we could use these operators to do model Quantization on our custom device.

do you have to use that? does https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/fx/_decomposed.py#L36 work for you?

**Comment by heidongxianhua** [comment]
> just FYI, new versions of pytorch quantization is planning to move away from QTensor and just use plain pytorch Tensor instead so that we are not tied to pytorch core: https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/fx/_decomposed.py#L56

But now, we could also call these by `torch.quantize_per_tensor`. And if we  are planning to move away from QTensor, when we refactor `QTensor`, use tensor instead in C++? or do we have a plan?

**Comment by heidongxianhua** [comment]
> > > just FYI, new versions of pytorch quantization is planning to move away from QTensor and just use plain pytorch Tensor instead so that we are not tied to pytorch core: https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/fx/_decomposed.py#L56
> > 
> > 
> > As for QTensor, there are many dedicated operators such as `quantize_per_tensor` and `quantize_per_tensor_dynamic`, in these operator, pytorch will create `QTensor`. we could use these operators to do model Quantization on our custom device.
> 
> do you have to use that? does https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/fx/_decomposed.py#L36 work for you?

I have tested, the APIs defined in _decomposed.py have different arguments with the same named APIs defined in `native_functions.yaml`, so it is maybe troublesome to use this with `_decomposed.py`. And another, if we are planning to move away from QTensor, does this mean that we will discard these operators defined in `native_functions.yaml` (such as `quantize_per_tensor`, `quantize_per_channel`, `dequantize` and some on, related to `dequantize` and `quantize`) and suggest to use the APIs defined in `_decomposed.py`? @jerryzh168

**Comment by ezyang** [comment]
Hmm, if you only want to override the operators like `quantize_per_tensor`, you can just register your custom kernel to that operator and override it entirely. This seems like the best thing to do to me. You don't need these hooks in that case.

**Pull request closed without merging**